File: cli/commands/business.ts
import {
  getBusinessInfo,
  listBusinessAdAccounts,
  listInvoices,
  downloadInvoicePdf,
  searchAdLibrary,
} from "../../src/domains/business.js";
import { graphApi } from "../../src/api.js";
import type { RuntimeContext } from "../lib/context.js";

const deps = { graphApi };

function requireToken(runtime: RuntimeContext): string {
  if (!runtime.accessToken) throw new Error("Missing access token. Use --access-token or profile/env token.");
  return runtime.accessToken;
}

export async function handleBusinessCommand(args: string[], runtime: RuntimeContext): Promise<unknown> {
  const [sub, ...rest] = args;
  if (!sub) throw new Error("Usage: fbcli business <info|ad-accounts> ...");
  if (runtime.dryRun) return { ok: true, route: `business ${sub}` };
  const token = requireToken(runtime);

  if (sub === "info") return getBusinessInfo(deps, rest[0] ?? "", token);
  if (sub === "ad-accounts") return listBusinessAdAccounts(deps, rest[0] ?? "", token);
  throw new Error("Usage: fbcli business <info|ad-accounts> <business-id>");
}

export async function handleInvoicesCommand(args: string[], runtime: RuntimeContext): Promise<unknown> {
  const [sub, ...rest] = args;
  if (!sub) throw new Error("Usage: fbcli invoices <list|download> ...");
  if (runtime.dryRun) return { ok: true, route: `invoices ${sub}` };
  const token = requireToken(runtime);

  if (sub === "list") {
    const [businessId, since, until] = rest;
    return listInvoices(deps, businessId ?? "", token, since, until);
  }
  if (sub === "download") {
    return {
      bytes: (await downloadInvoicePdf(deps, rest[0] ?? "", token)).byteLength,
      invoiceId: rest[0] ?? "",
    };
  }
  throw new Error("Usage: fbcli invoices <list|download> ...");
}

export async function handleAdLibraryCommand(args: string[], runtime: RuntimeContext): Promise<unknown> {
  const [sub, ...rest] = args;
  if (sub !== "search") throw new Error("Usage: fbcli ad-library search <query>");
  if (runtime.dryRun) return { ok: true, route: "ad-library search" };
  const token = requireToken(runtime);
  return searchAdLibrary(deps, token, { search_terms: rest[0] ?? "", ad_reached_countries: "US" });
}

File: internal/cmd/business.go
package cmd

import (
	"errors"

	"github.com/spf13/cobra"

	"github.com/salmonumbrella/facebook-cli/internal/facebook"
)

func newBusinessCommand(a *app) *cobra.Command {
	return &cobra.Command{
		Use:   "business",
		Args:  cobra.ArbitraryArgs,
		Short: "Business API commands",
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) < 1 {
				return errors.New("usage: fbcli business <info|ad-accounts>")
			}
			sub := args[0]
			if a.runtime.DryRun {
				return a.write(cmd, map[string]any{"ok": true, "route": "business " + sub})
			}
			token, err := a.requireAccessToken()
			if err != nil {
				return err
			}
			var result any
			switch sub {
			case "info":
				result, err = facebook.GetBusinessInfo(cmd.Context(), a.client, argAt(args, 1), token, nil)
			case "ad-accounts":
				result, err = facebook.ListBusinessAdAccounts(cmd.Context(), a.client, argAt(args, 1), token, nil)
			default:
				return errors.New("usage: fbcli business <info|ad-accounts> <business-id>")
			}
			if err != nil {
				return err
			}
			return a.write(cmd, result)
		},
	}
}

func newInvoicesCommand(a *app) *cobra.Command {
	return &cobra.Command{
		Use:   "invoices",
		Args:  cobra.ArbitraryArgs,
		Short: "Invoice commands",
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) == 0 {
				return errors.New("usage: fbcli invoices <list|download>")
			}
			sub := args[0]
			if a.runtime.DryRun {
				return a.write(cmd, map[string]any{"ok": true, "route": "invoices " + sub})
			}
			token, err := a.requireAccessToken()
			if err != nil {
				return err
			}
			switch sub {
			case "list":
				result, err := facebook.ListInvoices(cmd.Context(), a.client, argAt(args, 1), token, argAt(args, 2), argAt(args, 3), nil)
				if err != nil {
					return err
				}
				return a.write(cmd, result)
			case "download":
				bytes, err := facebook.DownloadInvoicePDF(cmd.Context(), a.client, argAt(args, 1), token)
				if err != nil {
					return err
				}
				return a.write(cmd, map[string]any{"bytes": len(bytes), "invoiceId": argAt(args, 1)})
			default:
				return errors.New("usage: fbcli invoices <list|download>")
			}
		},
	}
}

func newAdLibraryCommand(a *app) *cobra.Command {
	return &cobra.Command{
		Use:   "ad-library",
		Args:  cobra.ArbitraryArgs,
		Short: "Ad library search",
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) < 1 || args[0] != "search" {
				return errors.New("usage: fbcli ad-library search <query>")
			}
			if a.runtime.DryRun {
				return a.write(cmd, map[string]any{"ok": true, "route": "ad-library search"})
			}
			token, err := a.requireAccessToken()
			if err != nil {
				return err
			}
			result, err := facebook.SearchAdLibrary(cmd.Context(), a.client, token, map[string]string{
				"search_terms":         argAt(args, 1),
				"ad_reached_countries": "US",
			})
			if err != nil {
				return err
			}
			return a.write(cmd, result)
		},
	}
}
